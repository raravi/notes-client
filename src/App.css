* {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
}

body {
  font-size: 16px;
	line-height: 1.4em;
  font-weight: 300;
}

.container {
  max-width: 700px;
  min-width: 270px;
  margin: 0 auto;
  padding: 0 10px;
}

.header {
  padding: 50px 0;
  text-align: center;
}

.header__title {
  font-size: 60px;
  letter-spacing: 0.1em;
  -webkit-text-rendering: optimizeLegibility;
	-moz-text-rendering: optimizeLegibility;
	text-rendering: optimizeLegibility;
  color: darkgrey;
  font-family: monospace;
  font-weight: 200;
}

.note {
  min-height: 400px;
  height: 60vh;
  overflow-y: scroll;
  padding: 0 8px;
  border: 2px solid lightgrey;
  border-radius: 12px;
}

.note__line {
  display: block;
  width: 100%;
}

.note__paragraph {
  color: grey;
}

.note__header1 {
  color: grey;
  font-size: 28px;
  line-height: 2.5;
}

.note__header2 {
  color: grey;
  font-size: 26px;
  line-height: 2.5;
}

.note__header3 {
  color: grey;
  font-size: 24px;
  line-height: 2.5;
}

.note__header4 {
  color: grey;
  font-size: 22px;
  line-height: 2.5;
}

.note__header5 {
  color: grey;
  font-size: 20px;
  line-height: 2.5;
}

.note__header6 {
  color: grey;
  font-size: 18px;
  line-height: 2.5;
}

.footer {
  color: grey;
  font-size: 16px;
  letter-spacing: 0.02em;
  margin: 80px;
  text-align: center;
}

.footer a:visited {
  color: darkblue;
}
