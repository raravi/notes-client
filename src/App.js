import React from 'react';
import './App.css';

function App() {
  function setCaretPositionToOffset(el, offset) {
    el.focus();
    let range = document.createRange();
    let sel = window.getSelection();
    range.setStart(el.firstChild, offset);
    range.collapse(true);
    sel.removeAllRanges();
    sel.addRange(range);
  }

  function setCaretPositionInChildNode(node, offset) {
    let offsetLeft = offset;
    let children = node.childNodes;
    for (let i = 0; i < children.length; i++) {
      if (offsetLeft <= children[i].innerText.length) {
        setCaretPositionToOffset(children[i], offsetLeft);
        return;
      } else {
        offsetLeft -= children[i].innerText.length;
      }
    }
  }

  function checkForPairs(el, newText, length, offset, e) {
    // Not paragraph
    let text = newText.slice(length);
    console.log("text : '" + text + "'");
    let countOfStars = (text.match(new RegExp("\\*", "g")) || []).length;
    console.log("Count: " + countOfStars);
    if (countOfStars >= 1) {
      // Atleast one pair of matching 'markers'
      let indexOfStars = [];
      [...text].forEach((value, index) => {
        if (value === "*") {
          indexOfStars.push(length + index);
        }
      });
      if (countOfStars % 2 !== 0)
        indexOfStars.pop();
      console.log("indexOfStars : " + indexOfStars);
      // Add new child spans
      let newSpanText = "<span class='note__text'>";
      //let boldEnded = false;
      [...newText].forEach((value, index, array) => {
        if (index < length) {
          newSpanText += value;
        } else {
          let indexOfStarsIndex = indexOfStars.indexOf(index);
          if (value === "*" && indexOfStarsIndex !== -1) {
            if (indexOfStarsIndex % 2 === 0) {
              //newSpanText += "<span class='note__star'>" + value;
              newSpanText += "</span><span class='note__bold'>" + value;
            } else {
              newSpanText += value + "</span><span class='note__text'>";
              //boldEnded = true;
            }
          } else {
            newSpanText += value;
          }
        }
      });
      newSpanText += "\xa0</span>";
      console.log("New Span: '" + newSpanText + "'");
      console.log("Offset: ", offset);
      let pNode = el.parentNode;
      el.parentNode.innerHTML = newSpanText;
      setCaretPositionInChildNode(pNode, offset);
      e.preventDefault();
    }
  }

  function checkHeader(el, updatedText, offset, e) {
    let currentText;
    if (updatedText) {
      currentText = updatedText;
    } else {
      currentText = el.parentNode.innerText;
    }
    let newText = "";
    for (let i = 0; i < currentText.length-1; i++) {
      if (currentText[i] === '\xa0')
        newText += " ";
      else
        newText += currentText[i];
    }

    let strings = newText.split(" ");
    let firstString = strings[0];
    let regexForOrderedList = new RegExp('^\\d+\\.$');
    if (strings.length > 1 && firstString === "#") {
      console.log("Header 1");
      el.parentNode.setAttribute("class", "note__header1");
    } else if (strings.length > 1 && firstString === "##") {
      console.log("Header 2");
      el.parentNode.setAttribute("class", "note__header2");
    } else if (strings.length > 1 && firstString === "###") {
      console.log("Header 3");
      el.parentNode.setAttribute("class", "note__header3");
    } else if (strings.length > 1 && firstString === "####") {
      console.log("Header 4");
      el.parentNode.setAttribute("class", "note__header4");
    } else if (strings.length > 1 && firstString === "#####") {
      console.log("Header 5");
      el.parentNode.setAttribute("class", "note__header5");
    } else if (strings.length > 1 && firstString === "######") {
      console.log("Header 6");
      el.parentNode.setAttribute("class", "note__header6");
    } else if (strings.length > 1 && firstString === "*") {
      console.log("Unordered List");
      el.parentNode.setAttribute("class", "note__unorderedlist");
    } else if (strings.length > 1 && regexForOrderedList.test(firstString)) {
      console.log("Ordered List");
      el.parentNode.setAttribute("class", "note__orderedlist");
    } else if (strings.length > 1 && firstString === ">") {
      console.log("Block Quote");
      el.parentNode.setAttribute("class", "note__blockquote");
    } else {
      console.log("Para");
      el.parentNode.setAttribute("class", "note__paragraph");
    }

    // Handle *word*
    if (strings.length > 1 && ( firstString === "#" ||
                                firstString === "##" ||
                                firstString === "###" ||
                                firstString === "####" ||
                                firstString === "#####" ||
                                firstString === "######" ||
                                firstString === "*" ||
                                regexForOrderedList.test(firstString) ||
                                firstString === ">")) {
      // Not a paragraph
      checkForPairs(el, newText, firstString.length, offset, e);
    } else {
      // Paragraph
      checkForPairs(el, newText, 0, offset, e);
    }
  }

  function onKeyDownInEditor(e) {
    // Get current SPAN details
    let currentSelection = window.getSelection();
    let currentNode = currentSelection.anchorNode.parentNode;
    let currentOffset = currentSelection.anchorOffset;
    let currentText = currentNode.innerText;
    let parentOffset = 0;

    let children = currentNode.parentNode.childNodes;
    for (let i = 0; i < children.length; i++) {
      if (children[i] === currentNode) {
        parentOffset += currentOffset;
        break;
      } else {
        parentOffset += children[i].innerText.length;
      }
    }

    // trap the return key being pressed
    if (e.keyCode === 13) {
      // Prevent the <div /> creation.
      e.preventDefault();

      // Create new DIV with an 'empty' SPAN inside it
      let innerSpanElement = document.createElement('span');
      innerSpanElement.setAttribute("class", "note__text");
      innerSpanElement.innerText = '\xa0';

      let spanElement = document.createElement('span');
      spanElement.setAttribute("class", "note__paragraph");
      spanElement.appendChild(innerSpanElement);

      let divElement = document.createElement('div');
      divElement.setAttribute("class", "note__line");
      divElement.appendChild(spanElement);

      // Check for conditions to edit DIVs
      if (parentOffset === 0) {
        currentNode.parentNode.parentNode.parentNode.insertBefore(divElement, currentNode.parentNode.parentNode);
        // Set the Cursor position
        setCaretPositionToOffset(currentNode, 0);
      } else if (parentOffset === currentNode.parentNode.innerText.length-1) {
        // Insert new DIV after updating it above
        currentNode.parentNode.parentNode.parentNode.insertBefore(divElement, currentNode.parentNode.parentNode.nextSibling);
        // Set the Cursor position
        setCaretPositionToOffset(innerSpanElement, 0);
      } else {
        let indexOfSpan = -1, haveToBreakSpan, breakAtOffset = -1;
        let offsetLeft = parentOffset;
        let children = currentNode.parentNode.childNodes;
        for (let i = 0; i < children.length; i++) {
          if (offsetLeft < children[i].innerText.length) {
            // Have to break current SPAN
            indexOfSpan = i;
            haveToBreakSpan = true;
            breakAtOffset = offsetLeft;
            break;
          } else if (offsetLeft === children[i].innerText.length) {
            indexOfSpan = i;
            haveToBreakSpan = false;
            break;
          } else {
            offsetLeft -= children[i].innerText.length;
          }
        }
        // console.log("indexOfSpan :", indexOfSpan, "haveToBreakSpan :", haveToBreakSpan, "breakAtOffset :", breakAtOffset);
        let tempNodes = [];
        for (let i = 0; i < children.length; i++) {
          if (i === indexOfSpan) {
            if (haveToBreakSpan) {
              let tempText = children[i].innerText;
              children[i].innerText = tempText.slice(0, breakAtOffset) + '\xa0';

              let textElement = document.createElement('span');
              textElement.setAttribute("class", children[i].getAttribute("class"));
              textElement.innerText = tempText.slice(breakAtOffset);
              tempNodes.push(textElement);
            } else {
              children[i].innerText += '\xa0';
            }
          }
          if (i > indexOfSpan) {
            tempNodes.push(children[i]);
          }
        }
        spanElement.firstChild.remove();
        tempNodes.forEach((childNode) => {
          childNode.remove();
          spanElement.appendChild(childNode);
        });
        currentNode.parentNode.parentNode.parentNode.insertBefore(divElement, currentNode.parentNode.parentNode.nextSibling);
        // Check if Header or Paragraph
        checkHeader(currentNode, null, parentOffset, e);
        checkHeader(spanElement.firstChild, null, parentOffset, e);
        // Set the Cursor position
        setCaretPositionToOffset(spanElement.firstChild, 0);
      }

    } else if (e.key.length === 1) {
      // 'e.key' of length 1 represents the character to insert
      console.log("In length=1", e.key, e.keyCode);

      // Calculate new Text after addition of character from 'e.key'
      let stringBeforeCaret = currentNode.parentNode.innerText.slice(0, parentOffset);
      let stringAfterCaret = currentNode.parentNode.innerText.slice(parentOffset);
      currentText = "" + stringBeforeCaret + e.key + stringAfterCaret;

      // Check if Header or Paragraph
      checkHeader(currentNode, currentText, stringBeforeCaret.length+1, e);

    } else if (e.key === "Backspace") {
      console.log("In Backspace", e.key, e.keyCode);

      if (parentOffset === 0) {
        let previousDivNode = currentNode.parentNode.parentNode.previousSibling;
        if (previousDivNode && previousDivNode.getAttribute("class") === "note__line") {
          e.preventDefault(); // Prevent move to last cursor position.
          let previousOffset = previousDivNode.firstChild.innerText.length-1;
          let previousLastChild = previousDivNode.firstChild.lastChild;
          let previousText = previousLastChild.innerText;
          previousDivNode.firstChild.lastChild.innerText = previousText.slice(0, -1);
          previousDivNode.firstChild.innerHTML += currentNode.parentNode.innerHTML;

          setCaretPositionInChildNode(previousDivNode.firstChild, previousOffset);
          currentNode.parentNode.parentNode.remove();
        }
      } else {
        let stringBeforeCaret = currentNode.parentNode.innerText.slice(0, parentOffset);
        let stringAfterCaret = currentNode.parentNode.innerText.slice(parentOffset);
        currentText = "" + stringBeforeCaret.slice(0, -1) + stringAfterCaret;

        checkHeader(currentNode, currentText, stringBeforeCaret.length-1, e);
      }

    } else if (e.key === "ArrowRight") {
      console.log("In ArrowRight", e.key, e.keyCode);

      if (currentNode === currentNode.parentNode.lastChild && currentOffset === currentText.length-1) {
        e.preventDefault(); // Prevent the cursor move to right.
        let nextDivNode = currentNode.parentNode.parentNode.nextSibling;
        if (nextDivNode && nextDivNode.getAttribute("class") === "note__line") {
          setCaretPositionToOffset(nextDivNode.firstChild.firstChild, 0);
        }
      }

    } else if (e.key === "ArrowDown") {
      console.log("In ArrowDown", e.key, e.keyCode);

      let nextDivNode = currentNode.parentNode.parentNode.nextSibling;
      if (!nextDivNode) {
        e.preventDefault(); // Dont move to the last cursor position.
        setCaretPositionToOffset(currentNode.parentNode.lastChild, currentNode.parentNode.lastChild.innerText.length-1);
      } else if (nextDivNode.getAttribute("class") === "note__line") {
        e.preventDefault(); // Prevent the <div /> creation.
        let nextText = nextDivNode.firstChild.innerText;
        if (parentOffset >= nextText.length-1) {
          setCaretPositionToOffset(nextDivNode.firstChild.lastChild, nextDivNode.firstChild.lastChild.innerText.length-1);
        } else {
          setCaretPositionInChildNode(nextDivNode.firstChild, parentOffset);
        }
      }

    } else if (e.key === "ArrowLeft") {
      console.log("In ArrowLeft", e.key, e.keyCode);

      if (currentOffset === 0) {
        let previousDivNode = currentNode.parentNode.parentNode.previousSibling;
        if (previousDivNode && previousDivNode.getAttribute("class") === "note__line") {
          e.preventDefault(); // Prevent the <div /> creation.
          setCaretPositionToOffset(previousDivNode.firstChild.lastChild, previousDivNode.firstChild.lastChild.innerText.length-1);
        }
      }

    } else if (e.key === "ArrowUp") {
      console.log("In ArrowUp", e.key, e.keyCode);

      let previousDivNode = currentNode.parentNode.parentNode.previousSibling;
      if (previousDivNode && previousDivNode.getAttribute("class") === "note__line") {
        e.preventDefault(); // Prevent the <div /> creation.
        let previousText = previousDivNode.firstChild.innerText;
        if (parentOffset >= previousText.length-1) {
          setCaretPositionToOffset(previousDivNode.firstChild.lastChild, previousDivNode.firstChild.lastChild.innerText.length-1);
        } else {
          setCaretPositionInChildNode(previousDivNode.firstChild, parentOffset);
        }
      }

    } else {
      console.log("In else", e.key, e.keyCode);
    }
  }

  function onClick(e) {
    console.log("On Click");
    // Get current SPAN details
    let currentSelection = window.getSelection();
    let currentNode = currentSelection.anchorNode;
    if (currentNode.nodeType === 3)
      currentNode = currentNode.parentNode;
    let currentOffset = currentSelection.anchorOffset;
    let currentText = currentNode.innerText;

    if (currentOffset > currentText.length-1) {
      e.preventDefault();
      setCaretPositionToOffset(currentNode, currentText.length-1);
    } else {
      setCaretPositionToOffset(currentNode, currentOffset);
    }
  }

  return (
    <div className="container">
      <header className="header">
        <h1 className="header__title">Notes</h1>
      </header>
      <div contentEditable="true" className="note" onKeyDown={onKeyDownInEditor} onMouseUp={onClick}>
        <div className="note__line">
          <span className="note__header1">
            <span className="note__text"># Header </span>
            <span className="note__text">One&nbsp;</span>
          </span>
        </div>
        <div className="note__line">
          <span className="note__paragraph">
            <span className="note__text">Simple </span>
            <span className="note__text">lineeeeeeeeee&nbsp;</span>
          </span>
        </div>
        <div className="note__line">
          <span className="note__header1">
            <span className="note__text"># Head</span>
            <span className="note__bold">*er O*</span>
            <span className="note__text">ne&nbsp;</span>
          </span>
        </div>
        <div className="note__line">
          <span className="note__paragraph">
            <span className="note__text">Sim</span>
            <span className="note__bold">*ple l*</span>
            <span className="note__text">ine&nbsp;</span>
          </span>
        </div>
      </div>
      <footer className="footer">
        <p>&#169; 2020 Amith Raravi - source code on <a href="/">Github</a></p>
      </footer>
    </div>
  );
}

export default App;
