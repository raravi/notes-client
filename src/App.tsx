import React, { useState } from 'react';
import axios from 'axios';
import {
  Switch,
  Route,
  Redirect
} from "react-router-dom";
import { Dashboard } from './components/Dashboard';
import { Login } from './components/Login';
import jwtDecode from 'jwt-decode';
import { loginApiDetails, appApiDetails } from './config/apiDetails';
import './App.css';

type Response = {
  data : {
    token: string;
  }
};

type TokenDecoded = {
  id: string,
  name: string,
};

/**
 * The main App component.
 * Routes to Login / Dashboard React Components based upon the LOGIN state.
 */
function App() {
  let [ userLoggedIn, setUserLoggedIn ] = useState<TokenDecoded | null>(null);
  let [ notes, setNotes ] = useState<Note[]>([]);

  function onSuccessfulLogin(response: Response) {
    if (response && response.data && response.data.token) {
      let tokenDecoded : TokenDecoded = jwtDecode(response.data.token);
      if(tokenDecoded) {
        sessionStorage.setItem("token", response.data.token);

        axios.post(appApiDetails.url + appApiDetails.endpoints.initialSync, {
          userid: tokenDecoded.id
        },
        {
          withCredentials: true,
          headers: {
            Authorization: sessionStorage.getItem("token")
          }
        })
        .then(responseSync => {
          if (responseSync && responseSync.data && responseSync.data.success) {
            setNotes(responseSync.data.notes);
            setUserLoggedIn({id: tokenDecoded.id, name: tokenDecoded.name});
          }
        })
        .catch(error => console.log(error));
      }/* else {
        loginDispatch({ type: 'email-error', text: "An error occured..." });
      }*/
    }
  }

  return (
    <Switch>
      <Route path="/dashboard">
        { userLoggedIn
          ? <Dashboard  setUserLoggedIn={setUserLoggedIn}
                        notes={notes}
                        setNotes={setNotes}
                        user={userLoggedIn} />
          : <Redirect to="/" />
        }
      </Route>
      <Route path="/">
        { userLoggedIn
          ? <Redirect to="/dashboard" />
          : <Login
              loginSuccessCallback={onSuccessfulLogin}
              apiDetails={loginApiDetails} />
        }
      </Route>
    </Switch>
  );
}

export default App;
