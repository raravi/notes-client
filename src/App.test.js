import React from 'react';
import { act, render, cleanup, fireEvent, waitForElement } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import axiosMock from 'axios';
import App from './App';

jest.mock('axios');

afterEach(cleanup);

it('should take a snapshot', () => {
  const { asFragment } = render(<App />);
  expect(asFragment(<App />)).toMatchSnapshot();
});

/**
 * LOGIN page
 */
describe('Login Page', () => {
  it('submit button is present', () => {
    const { getByTestId } = render(<App />);

    expect(getByTestId('login-button')).toBeInTheDocument();
  });

  it('login is successful', async () => {
    axiosMock.post.mockResolvedValueOnce({
      data: {
        success: true,
        token: 'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6IjVlNWVkY2E0M2FhOWRjNTg3NTAzZTFiNCIsIm5hbWUiOiJBbWl0aCBSYXJhdmkiLCJpYXQiOjE1ODQ0MDQxNzQsImV4cCI6MTYxNTk2MTEwMH0.qLqhN_1CHbQLOCXzOyxRZS9K42AsoHtQbF-qL8vgn0o',
        notes: []
      }
    });
    const { getByTestId, findByTestId } = render(<App />);

    fireEvent.click(getByTestId('login-button'));

    const dashboardNoteElement = await findByTestId('dashboard-note');
    expect(axiosMock.post).toHaveBeenCalledTimes(1);
    expect(axiosMock.post).toHaveBeenCalledWith("http://localhost:8000/api/users/login", {"email": "", "password": ""});
  });

  it('login error occured', async () => {
    axiosMock.post.mockImplementation(() => Promise.reject({
      response: {
        status: 404,
        data: {
          email: "Email not found",
          password: "Password incorrect"
        }
      }
    }));
    const { getByTestId, findByText, findByTestId } = render(<App />);

    fireEvent.click(getByTestId('login-button'));

    const emailErrorElement = await findByText('Email not found');
    const passwordErrorElement = await findByText('Password incorrect');
    expect(axiosMock.post).toHaveBeenCalledTimes(2);
    expect(axiosMock.post).toHaveBeenCalledWith("http://localhost:8000/api/users/login", {"email": "", "password": ""});
  });

  it('Register link is clicked', () => {
    const { getByTestId } = render(<App />);

    fireEvent.click(getByTestId('login-register'));

    expect(getByTestId('register-button')).toBeInTheDocument();
  });

  it('Forgot Password is clicked', () => {
    const { getByTestId } = render(<App />);

    fireEvent.click(getByTestId('login-forgotpassword'));

    expect(getByTestId('forgotpassword-button')).toBeInTheDocument();
  });
});

/**
 * REGISTER page
 */
describe('Register Page', () => {
  it('submit button is present', () => {
    const { getByTestId } = render(<App />);

    fireEvent.click(getByTestId('login-register'));

    expect(getByTestId('register-button')).toBeInTheDocument();
  });

  it('Register is successful', async () => {
    axiosMock.post.mockResolvedValueOnce({
      data: {createduser: "New user registered successfully!"}
    });
    const { getByTestId, findByTestId, findByText } = render(<App />);
    fireEvent.click(getByTestId('login-register'));
    expect(getByTestId('register-button')).toBeInTheDocument();

    fireEvent.click(getByTestId('register-button'));

    const registerSuccessElement = await findByText('New user registered successfully!');
    expect(axiosMock.post).toHaveBeenCalledTimes(3);
    expect(axiosMock.post).toHaveBeenCalledWith("http://localhost:8000/api/users/register", {"name": "", "email": "", "password": "", "password2": ""});
  });

  it('Register error occured', async () => {
    axiosMock.post.mockImplementation(() => Promise.reject({
      response: {
        status: 404,
        data: {
          name: "Name field is required",
          email: "Email already exists",
          password: "Password must be at least 6 characters",
          password2: "Passwords must match"
        }
      }
    }));
    const { getByTestId, findByTestId, findByText } = render(<App />);
    fireEvent.click(getByTestId('login-register'));
    expect(getByTestId('register-button')).toBeInTheDocument();

    fireEvent.click(getByTestId('register-button'));

    const nameErrorElement = await findByText('Name field is required');
    const emailErrorElement = await findByText('Email already exists');
    const passwordErrorElement = await findByText('Password must be at least 6 characters');
    const password2ErrorElement = await findByText('Passwords must match');
    expect(axiosMock.post).toHaveBeenCalledTimes(4);
    expect(axiosMock.post).toHaveBeenCalledWith("http://localhost:8000/api/users/register", {"name": "", "email": "", "password": "", "password2": ""});
  });

  it('Login link is clicked', () => {
    const { getByTestId } = render(<App />);
    fireEvent.click(getByTestId('login-register'));
    expect(getByTestId('register-button')).toBeInTheDocument();

    fireEvent.click(getByTestId('register-login'));

    expect(getByTestId('login-button')).toBeInTheDocument();
  });
});

/**
 * FORGOT PASSWORD page
 */
describe('Forgot Password Page', () => {
  it('submit button is present', () => {
    const { getByTestId } = render(<App />);

    fireEvent.click(getByTestId('login-forgotpassword'));

    expect(getByTestId('forgotpassword-button')).toBeInTheDocument();
  });

  it('Forgot Password is successful', async () => {
    axiosMock.post.mockResolvedValueOnce({
      data: {emailsent: 'The reset email has been sent, please check your inbox!'}
    });
    const { getByTestId, findByTestId, findByText } = render(<App />);
    fireEvent.click(getByTestId('login-forgotpassword'));
    expect(getByTestId('forgotpassword-button')).toBeInTheDocument();

    fireEvent.click(getByTestId('forgotpassword-button'));

    const forgotPasswordSuccessElement = await findByText('The reset email has been sent, please check your inbox!');
    expect(axiosMock.post).toHaveBeenCalledTimes(5);
    expect(axiosMock.post).toHaveBeenCalledWith("http://localhost:8000/api/users/forgotpassword", {"email": ""});
  });

  it('Forgot Password error occured', async () => {
    axiosMock.post.mockImplementation(() => Promise.reject({
      response: {
        status: 404,
        data: { email: "Email not found" }
      }
    }));
    const { getByTestId, findByTestId, findByText } = render(<App />);
    fireEvent.click(getByTestId('login-forgotpassword'));
    expect(getByTestId('forgotpassword-button')).toBeInTheDocument();

    fireEvent.click(getByTestId('forgotpassword-button'));

    const forgotPasswordErrorElement = await findByText('Email not found');
    expect(axiosMock.post).toHaveBeenCalledTimes(6);
    expect(axiosMock.post).toHaveBeenCalledWith("http://localhost:8000/api/users/forgotpassword", {"email": ""});
  });
});

/**
 * DASHBOARD page
 */
describe('Dashboard Page', () => {
  it('dashboard note is present', async () => {
    axiosMock.post.mockResolvedValueOnce({
      data: {
        success: true,
        token: 'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6IjVlNWVkY2E0M2FhOWRjNTg3NTAzZTFiNCIsIm5hbWUiOiJBbWl0aCBSYXJhdmkiLCJpYXQiOjE1ODQ0MDQxNzQsImV4cCI6MTYxNTk2MTEwMH0.qLqhN_1CHbQLOCXzOyxRZS9K42AsoHtQbF-qL8vgn0o',
        notes: []
      }
    });
    const { getByTestId, findByTestId } = render(<App />);

    fireEvent.click(getByTestId('login-button'));

    const dashboardNoteElement = await findByTestId('dashboard-note');
  });
});
