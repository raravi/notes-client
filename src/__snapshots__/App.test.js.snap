// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="login-container"
  >
    <header
      class="login-header"
    >
      <h1
        class="login-header__title"
      >
        notes
      </h1>
    </header>
    <section
      class="login"
    >
      <div
        class="login__header-group"
      >
        <h2
          class="login__header login__header--selected"
        >
          Login
        </h2>
        <h2
          class="login__header"
          data-testid="login-register"
        >
          Register
        </h2>
      </div>
      <input
        class="login__email"
        name="email"
        placeholder="Email"
        type="email"
      />
      <br />
      <label
        class="login__email-error"
        data-testid="login-email-error"
      />
      <input
        class="login__password"
        name="password"
        placeholder="Password"
        type="password"
      />
      <label
        class="login__password-error"
      />
      <br />
      <button
        class="login__submit"
        data-testid="login-button"
        type="button"
      >
        Submit
      </button>
      <div
        class="login__forgot-passwordblock"
      >
        <span
          class="login__forgot-password"
          data-testid="login-forgotpassword"
        >
          Forgot Password
        </span>
      </div>
    </section>
    <footer
      class="login-footer"
    >
      <p>
        Â© 2020 Amith Raravi - source code on 
        <a
          href="https://github.com/raravi/notes-client"
        >
          Github
        </a>
      </p>
    </footer>
  </div>
</DocumentFragment>
`;
